Add EC2 metadata and OS environment variable expansion support in the config

trav-c opened the pull request:
This branch adds support for variable expansion in the config file, in particular I had a need to name the log stream with general category information, but still include unique instance information (in this case the id) as part of the name on auto-scaling instances. With this patch that can now be done directly in the config using eg:
`log_stream = "UAT-WEB-${InstanceID}"`

Commit:
Add EC2 metadata and OS environment variable expansion in the config file

Commit:
Remove unused reflection variable

apparentlymart commented:
Thanks for this, @trav-c! This is a great idea.

I am a little concerned with mixing the two variable types in the same namespace here, since it could be confusing if e.g. you typo a name and don't get an error because it tried to use an environment variable instead.

I wonder if this could use hashicorp/hil (which is designed to work with HCL, the config language we are using here) and support interpolations like `${env.HOME}` and `${instance.id}`.

I realize this is asking for a rewrite of a lot of what you did here so l don't expect that you must do it (though if you have the time, that would be awesome!) but it does seem like if I were to merge this now it would be a breaking change to add HIL later, so I'd like to pause here for now and consider that path first.

trav-c commented:
I can re-look into HIL as an option, I did in fact look at it briefly but struggled to find much documentation or simple examples for just value expansion.

As an interim option I think ${env.<Name>} and ${instance.<NAME>} can be supported fairly easily using os.Expand() so as to be forward compatible with any future implementation using HIL

apparentlymart commented:
@trav-c supporting the prefixes via the simpler mechanism does sound like a good compromise. HIL is indeed rather more complex, and probably not worth the effort right now as long as there's no use-case for functions and expressions.

Commit:
Change config variable expansion to the form ${instance.<key>} and ${env.<key>}
to be forwards compatible with a possible future enhancement using hashicorp/hil
Also switches os.Expand() to a private variant which only expands ${name} and not
$name to minimise impact to any existing configs that may contain $

trav-c commented:
@apparentlymart  I've updated the implementation to use ${instance.xxxx} and ${env.xxxx} including switching to a private modified version of os.Expand() that will leave $xxxx alone (ie only ${xxxx} is expanded now).

I have kept the ${instance.xxxx} xxxx key naming as per the InstanceIdentityDocument though, because doing it through reflection rather than mapping them manually means that any future keys will automatically be supported straight away.

apparentlymart commented:
Awesome! Thanks for working on this, and taking the time to iterate on it.
